addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js

```diff
@@ -3903,16 +3903,17 @@ export class OdooEditor extends EventTarget {
         this._pluginCall('cleanForSave', [element]);
 
         // Clean the zero-width spaces added by the `fillEmpty` function
-        // (flagged with the "data-oe-zws-empty-inline" attributes).
-        // Reverse the list to start from the deepest elements (for emptiness checks).
-        for (const emptyElement of element.querySelectorAll('[data-oe-zws-empty-inline]')) {
-            if (isZWS(emptyElement)) {
+        // (flagged with the "data-oe-zws-empty-inline" attributes). Reverse the
+        // list to start from the deepest elements (for emptiness checks).
+        for (const emptyElement of [...element.querySelectorAll('[data-oe-zws-empty-inline]')].reverse()) {
+            // if ([...emptyElement.textContent].every(char => char === '\u200B')) {
```

Comment:
Just FYI, the comment is my former implementation of this (before `isZWS` was introduced). It did one extra thing compared to the new version, which is that it also worked if the text was for instance `\u200B\u200B`. Or `          \u200B          `. And I'm just not 100% sure we don't want to handle that. Either way, `isZWS` is clearly insufficient.

Reply:
But wouldn't the node fall into the `else` clause then and go through `cleanZWS` ? I'm not sure I understand the issue here.

Reply:
Well yeah but the `else` doesn't do the same thing as the `if`. Not sure I get your comment ^^'

Examples of differences:
1. `<span data-oe-zws-empty-inline class="yo">\u200B\u200B</span>`
- With check 1 (`isZWS(emptyElement) || emptyElement.textContent.trim() === ''`), we go into the `else` -> remove the zws and the data attribute -> `<span></span>`.
- With check 2 (`[...emptyElement.textContent].every(char => char === '\u200B')`), we go into the `if` -> there's a class so we just remove the data attribute -> `<span class="yo">\u200B\u200B</span>`
2. `<span data-oe-zws-empty-inline>\u200B\u200B</span>`
- Check 1 -> `else` -> `<span></span>`
- Check 2 -> `if` -> no class -> element is removed
3. `<span data-oe-zws-empty-inline>    \u200B    </span>`
- Check 1 -> `else` -> `<span>        </span>`
- Check 2 (modified: add `.trim()` after `.textContent`) -> `if` -> no class -> element is removed.

Reply:
Alright, I see. Shouldn't the whole thing be this then ?
```js
cleanZWS(emptyElement);
emptyElement.removeAttribute('data-oe-zws-empty-inline');
if (emptyElement.textContent.trim() === '') {
    if (!emptyElement.classList.length) {
            // We only remove the empty element if it has no class, to
            // ensure we don't break visual styles (in that case, its
            // ZWS was kept to ensure the cursor can be placed in it).
            emptyElement.remove();
        }
}
```

Reply:
Yes I think that would have the same effect.

Reply:
@Zinston According to runbot it doesn't have the same effect... ^^'

Reply:
@dmo-odoo Ah no of course :facepalm: In the case where we enter the if and there's no class, we normally remove the data attribute without removing the zws, but with your version we remove it anyway.

This is for cases like this test shows:
1. `<p>ab<span class="style">[]cd</span>ef</p>`
2. `deleteForward` twice
3. `<p>ab<span class="style">[]\u200B</span>ef</p>`
With your version there's no zws at the end so we can't have the cursor there.

Reply:
Does it make sense to remove the data attribute if we don't remove the zws ? How is it going to be cleaned afterwards ? AFAIK we don't clean user-generated ZWS so we need to be able to identify the ones we created for technical reason, which is the point of the data attribute.

Reply:
I guess you're right, it's pretty weird... I get that we don't want to save that data attribute, but then we should actually remove that whole span anyway because we don't want to save it either. Either way, if we keep the zws, we should keep the data attribute.
Shouldn't the correct answer in this test be `contentAfterEdit: '<p>ab<span class="style" data-oe-zws-empty-inline>[]\u200B</span>ef</p>', contentAfter: '<p>ab[]ef</p>'`?

Reply:
Ah no apparently we do want to keep the span even if it's empty, I guess in case it has like a padding and a bgcolor or something...

![image](https://github.com/odoo/odoo/assets/24205914/b5c72f29-cb33-48d2-a3c1-eed88dc5c564)

Reply:
I think it would make more sense that way, yes.

Reply:
Ah yeah right, because it has a class. And then in that case we don't want to touch it as we don't know what it is supposed to do in the DOM. We only remove our special attribute because we know it can't be used for anything else.

So maybe this ?
```js
emptyElement.removeAttribute('data-oe-zws-empty-inline');
if (!emptyElement.classList.length) {
    cleanZWS(emptyElement);
    if (emptyElement.textContent.trim() === '') {
        // We only remove the empty element if it has no class, to
        // ensure we don't break visual styles (in that case, its
        // ZWS was kept to ensure the cursor can be placed in it).
        emptyElement.remove();
    }
}
```

Reply:
Ah, well, no. If the element has text, we should remove its zws even if it has a class.

I'm starting to think my original approach was best.

If my `if ([...emptyElement.textContent.trim()].every(char => char === '\u200B')) {` offends you, we can probably do `if (/^[\s\u200b]*$/.test(emptyElement.textContent)) {` instead.

Reply:
Alright but don't we already have a regexp like this one somewhere ? It feels familiar.

Reply:
I also thought it felt familiar but I think that's because of `const nonWhitespacesRegex = /[\S\u00A0]/;`.